pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/kubewharf/katalyst-core/pkg/util/external/mbm"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+)
+
+const (
+	MemoryBandwidthManagement = "mbm"
+
+	// controlling hype-parameters in percentage
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_PAINPOINT       = 105
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_SWEETPOINT      = 95
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_UNTHROTTLEPOINT = 80
+
+	MEMORY_BANDWIDTH_INCREASE_MIN = 256 // the minimum incremental value
+	MEMORY_BANDWIDTH_DECREASE_MIN = 512 // the minimum decremental value
+)
+
+type NUMAStater interface {
+	GetMachineState() state.NUMANodeMap
+}
+
+type Controller struct {
+	metricEmitter metrics.MetricEmitter
+	metricReader  types.MetricsReader
+	numaStater    NUMAStater
+	mbAdjust      mbm.MBAdjuster
+
+	numaThrottled sets.Int
+
+	packageMap         map[int][]int // package id --> numa nodes in the package
+	interval           time.Duration
+	bandwidthThreshold int64
+	minDeductionStep   float64
+	minIncreaseStep    float64
+}
+
+func (c Controller) Run(ctx context.Context) {
+	general.Infof("mbm controller is starting")
+	wait.Until(c.run, c.interval, ctx.Done())
+}
+
+func (c Controller) run() {
+	for p, nodes := range c.packageMap {
+		c.processPackage(p, nodes)
+	}
+}
+
+// getActiveNodeSets divides the nodes into workload-associated sets
+// only nodes having active workload assigned shall be considered
+func (c Controller) getActiveNodeSets(nodes []int) map[string]sets.Int {
+	nodeSets := make(map[string]sets.Int)
+	states := c.numaStater.GetMachineState()
+	for _, node := range nodes {
+		nodeState := states[node]
+		for podUID := range nodeState.PodEntries {
+			if _, ok := nodeSets[podUID]; !ok {
+				nodeSets[podUID] = sets.Int{}
+			}
+			nodeSets[podUID].Insert(node)
+		}
+	}
+	return nodeSets
+}
+
+// todo: to leverage the active group MB of whole package to save metric store fetching
+func (c Controller) getNodeMBMetrics(nodes sets.Int) (strategy.GroupMB, error) {
+	mbs := make(map[int]float64)
+	for node := range nodes {
+		nodeMB, err := c.metricReader.GetNumaMetric(node, "mb-mbps")
+		if err != nil {
+			return nil, err
+		}
+		mbs[node] = nodeMB.Value
+	}
+	return mbs, nil
+}
+
+func (c Controller) getActiveGroupMBs(nodes []int) (toSkip bool, mbs strategy.GroupMBs, err error) {
+	activeGroups := c.getActiveNodeSets(nodes)
+	if len(activeGroups) <= 1 {
+		// no need to take any action; neither an error
+		toSkip = true
+		return
+	}
+
+	mbs, err = c.getGroupMBMetrics(activeGroups)
+	return
+}
+
+func (c Controller) getGroupMBMetrics(actives map[string]sets.Int) (strategy.GroupMBs, error) {
+	groups := make(strategy.GroupMBs, 0)
+	for _, nodes := range actives {
+		groupMB, err := c.getNodeMBMetrics(nodes)
+		if err != nil {
+			return nil, err
+		}
+		groups = append(groups, groupMB)
+	}
+	return groups, nil
+}
+
+func (c Controller) processPackage(packageID int, nodes []int) {
+	// get the metrics
+	currMetric, err := c.metricReader.GetPackageMetric(packageID, "rw-mbps")
```

Comment:
define "rw-mbps" as consts in pkg/consts/metric.go

Reply:
sure; and will change other string literals in similar way.